### ISSUE TITLE
Disposing Watcher throws System.IO.IOException

### ISSUE THREAD

**Issue by torjue** [kubernetes-client/csharp#360]
When trying to dispose a `Watcher<T>`, using the following code and 1.6.12
```csharp
var result = await _client.ListClusterCustomObjectWithHttpMessagesAsync(
    "example.com",
    "v1",
    "someentries",
    watch: true,
    timeoutSeconds: (int)TimeSpan.FromMinutes(10).TotalSeconds,
    cancellationToken: cancelationToken);

var watcher = result.Watch(
    (WatchEventType type, SomeEntry item) => Console.WriteLine("{0} {1}", type, item.Metadata.Name),
    e => Console.WriteLine(e.ToString()),
    () => Console.WriteLine("Closed")
    );

await Task.Delay(5000);

watcher.Dispose();
```
the followng error is thrown:
```
System.IO.IOException: Unable to read data from the transport connection: Operation canceled.
---> System.Net.Sockets.SocketException (89): Operation canceled
    at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
    --- End of inner exception stack trace ---
    at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
    at System.Net.Security.SslStream.FillBufferAsync[TReadAdapter](TReadAdapter adapter, Int32 minSize)
    at System.Net.Security.SslStream.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory`1 buffer)
    at System.Net.Security.SslStream.Read(Byte[] buffer, Int32 offset, Int32 count)
    at System.Net.Http.HttpConnection.Fill()
    at System.Net.Http.HttpConnection.ChunkedEncodingReadStream.Read(Span`1 buffer)
    at System.Net.Http.HttpBaseStream.Read(Byte[] buffer, Int32 offset, Int32 count)
    at System.IO.StreamReader.ReadBuffer()
    at System.IO.StreamReader.ReadLine()
    at k8s.WatcherDelegatingHandler.PeekableStreamReader.ReadLine()
    at System.IO.TextReader.<>c.<ReadLineAsync>b__13_0(Object state)
    at System.Threading.Tasks.Task`1.InnerInvoke()
    at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
at k8s.Watcher`1.WatcherLoop(CancellationToken cancellationToken)
```

My guesses are that the underlying `_streamReader` is closed while the `while` loop is running, so `_streamReader.ReadLineAsync()` fails. Moving the `cancellationToken.IsCancellationRequested` into the `while` loop (as the first condition) might do the trick?

https://github.com/kubernetes-client/csharp/blob/e11cc58e56ca3d00a91177cd2ab45af547640dd9/src/KubernetesClient/Watcher.cs#L104-L109

**Comment by brendandburns** [comment]
We'd be happy to take a PR with this fix (and a test that validates it) but we'll also get to this eventually if you can't put a PR together.

**Comment by tg123** [comment]
it is by design

imagine that the code passed the first check you sugggest, `cancellationToken.IsCancellationRequested`, then runs into `ReadLineAsync`
you will still have this exception

**Comment by torjue** [comment]
Not sure the scenario you mention will happen? As `Dispose` canceles the `cancellationTokenSource` before it disposes `_streamReader`?

I'll see what I'm able to do regarding a PR.

**Comment by Woazboat** [comment]
I just ran into this issue as well and spent hours trying to figure out what I was doing wrong until I looked at the kubernetes client code and realized this is a problem with the implementation. 

As it is now, it is pretty much impossible to dispose the watcher without running into this major race condition. This makes the `OnError` event handler fairly useless (or at least very cumbersome to use) because the watch loop is pretty much guaranteed to throw an ObjectDisposedException when trying to read from the disposed stream reader.
